src/client/Microsoft.Identity.Client/Platforms/Android/AndroidBrokerHelper.cs

```diff
@@ -0,0 +1,488 @@
+using Microsoft.Identity.Client.Internal.Broker;
+using Microsoft.Identity.Client.Utils;
+using Microsoft.Identity.Json.Linq;
+
+namespace Microsoft.Identity.Client.Platforms.Android
+{
+    internal class AndroidBrokerHelper
+    {
+        private const string RedirectUriScheme = "msauth";
+        private const string BrokerTag = BrokerConstants.Signature;
+        public const string WorkAccount = "com.microsoft.workaccount.user.info";
+
+        private readonly Context _androidContext;
+        private readonly AccountManager _androidAccountManager;
+        private readonly ICoreLogger _logger;
+
+        public AndroidBrokerHelper(Context androidContext, ICoreLogger logger)
+        {
+            _androidContext = androidContext ?? throw new ArgumentNullException(nameof(androidContext));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            _logger.Verbose("Getting the Android context");
+            _androidAccountManager = AccountManager.Get(_androidContext);
+        }
+
+        public bool CanSwitchToBroker()
+        {
+            string packageName = _androidContext.PackageName;
+
+            // ADAL switches broker for following conditions:
+            // 1- app is not skipping the broker
+            // 2- signature of the broker is valid
+            // 3- account exists
+
+            //Force this to return true for broker test app
+
+            return VerifyAuthenticator(_androidAccountManager)
+                   && !packageName.Equals(BrokerConstants.PackageName, StringComparison.OrdinalIgnoreCase)
+                   && !packageName
+           .Equals(BrokerConstants.AzureAuthenticatorAppPackageName, StringComparison.OrdinalIgnoreCase);
+        }
+
+        public Intent GetIntentForInteractiveBrokerRequest(IDictionary<string, string> brokerPayload, Activity callerActivity)
+        {
+            Intent intent = null;
+
+            try
+            {
+                IAccountManagerFuture result = null;
+                // Callback is not passed since it is making a blocking call to get
+                // intent. Activity needs to be launched from calling app
+                // to get the calling app's metadata if needed at BrokerActivity.
+
+                Bundle addAccountOptions = new Bundle();
+                addAccountOptions.PutString(BrokerConstants.BrokerAccountManagerOperationKey, BrokerConstants.GetIntentForInteractiveRequest);
+
+                _logger.Info("Broker Account Name: " + GetValueFromBrokerPayload(brokerPayload, BrokerParameter.Username));
+
+                result = _androidAccountManager.AddAccount(BrokerConstants.BrokerAccountType,
+                    BrokerConstants.AuthtokenType, null, addAccountOptions, null,
+                    null, GetPreferredLooper(callerActivity));
+
+                if (result == null)
+                {
+                    _logger.Info("Android account manager AddAccount didn't return any results. ");
```

Comment:
can you specify broker somewhere  in here?

Reply:
Can you explain what you mean?

Reply:
yeah sure...like "Broker result is null. Android account manager AddAccount..." something like that, which specifies the call is from broker so when we look at the logs, it's clear where we are, without having to look in the code 

---
In reply to: [355480764](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1510#discussion_r355480764) [](ancestors = 355480764)